pkg/packetfilter/packetfilter.go

```diff
@@ -0,0 +1,273 @@
+	ChainTypeRoute            // mangle
+	ChainTypeNAT
+	ChainTypeMAX
+)
+
+type ChainPolicy uint32
+
+const (
+	ChainPolicyAccept ChainType = iota
+	ChainPolicyDrop
+)
+
+type Rule struct {
+	DestCIDR string
+	SrcCIDR  string
+
+	SrcSetName  string
+	DestSetName string
+
+	SnatCIDR string
+	DnatCIDR string
+
+	OutIntreface string
+	TargetChain  string
+	MssValue     string
+
+	DPort     string
+	MarkValue string
+	Action    RuleAction
+	Proto     RuleProto
+	ClampType MssClampType
+}
+
+// Supported policy values are accept (which is the default) or drop.
+type ChainRegular struct {
+	Name   string
+	Policy ChainPolicy
+}
+
+type ChainIPHook struct {
+	Name     string
+	Type     ChainType
+	Hook     ChainHook
+	Priority ChainPriority
+	Policy   ChainPolicy
+	JumpRule *Rule
+}
+
+type Base interface {
+	// Chains
+	ChainExists(table TableType, chain string) (bool, error)
+	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
+	CreateRegularChainIfNotExists(table TableType, chain *ChainRegular) error
+	DeleteIPHookChain(chain *ChainIPHook) error
+	DeleteRegularChain(table TableType, chain string) error
+	ClearChain(table TableType, chain string) error
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	InsertUnique(table TableType, chain string, position int, ruleSpec *Rule) error
+	PrependUnique(table TableType, chain string, ruleSpec *Rule) error
+	UpdateChainRules(table TableType, chain string, rules []*Rule) error
+	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
+}
+
+type MssClampInterface interface {
+	SetMultiClusterMssClamp(clampType MssClampType, mssValue int) error
+	DelMultiClusterMssClamp() error
+	AddSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+	DelSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+}
+type GlobalNetEgressInterface interface {
+	CreateGlobalEgressForPods(key, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForPods(handle GlobalEgressHandle) error
+	CreateGlobalEgressForNamespace(namespace, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForNamespace(handle GlobalEgressHandle) error
+	AddPodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	RemovePodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+}
+
+type Driver interface {
+	Base
+	MssClampInterface
+	GlobalNetEgressInterface
+}
+type DriverCreateFunc func() (Driver, error)
+
+type driverInfo struct {
+	createFunc DriverCreateFunc
+	name       string
+	isDefault  bool
+}
+
+var (
+	logger                 = log.Logger{Logger: logf.Log.WithName("PacketFilter")}
+	NewFunc                func() (Base, error)
+	NewMMssClampFunc       func() (MssClampInterface, error)
+	NewGlobalNetEgressFunc func() (GlobalNetEgressInterface, error)
+
+	drivers             []driverInfo
+	selectedDriverIndex int
+)
+
+func AddDriver(name string, isDefault bool, createFunc DriverCreateFunc) {
```

Comment:
What if multiple register as the default? Below it looks like it always takes the default, which I assume will be `iptables`. My understanding is that we'd dynamically determine which driver to use based on what's installed/supported, in which case the `DriverCreateFunc` would also return a bool indicating if it's supported (an err indicates it couldn't be determined). If both `iptables` and `nftables` can be installed simultaneously then we'd need a priority ranking to decide.

Reply:
I thought each driver should be responsible for determining if it is supported and register only if it is supported. 

Submariner should use the same packetfilter driver that cni and kube-proxy use, from what I understand cni and kube-proxy will use either iptables or nftables, so only one of iptables,nftables should be registered as a driver.

But since we still don't know how cni and kube-proxy will migrate to nftables I added the isDefault param. maybe I should delete it ?

Reply:
> I thought each driver should be responsible for determining if it is supported and register only if it is supported.

yeah we could do it that way but that might imply that only one driver should ever register, otherwise we'd still have to reconcile multiple registrations. We could just have _main.go_ figure out which driver to use and change this to `SetDriver` or, better yet, pass in the driver instance to `New`. Then there's be no ambiguity.

Reply:
Actually we wouldn't even need the `New` function.